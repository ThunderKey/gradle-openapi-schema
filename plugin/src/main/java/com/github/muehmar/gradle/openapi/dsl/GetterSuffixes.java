package com.github.muehmar.gradle.openapi.dsl;

import io.github.muehmar.pojobuilder.annotations.Nullable;
import io.github.muehmar.pojobuilder.annotations.PojoBuilder;
import java.io.Serializable;
import java.util.Optional;
import lombok.Data;
import lombok.With;

@Data
@With
@PojoBuilder
public class GetterSuffixes implements Serializable {
  private static final String DEFAULT_REQUIRED = "";
  private static final String DEFAULT_OPTIONAL = "Opt";
  private static final String DEFAULT_REQUIRED_NULLABLE = "Opt";
  private static final String DEFAULT_OPTIONAL_NULLABLE = "Tristate";

  @Nullable private String requiredSuffix;
  @Nullable private String requiredNullableSuffix;
  @Nullable private String optionalSuffix;
  @Nullable private String optionalNullableSuffix;

  public GetterSuffixes() {
    this(null, null, null, null);
  }

  public GetterSuffixes(
      String requiredSuffix,
      String requiredNullableSuffix,
      String optionalSuffix,
      String optionalNullableSuffix) {
    this.requiredSuffix = requiredSuffix;
    this.requiredNullableSuffix = requiredNullableSuffix;
    this.optionalSuffix = optionalSuffix;
    this.optionalNullableSuffix = optionalNullableSuffix;
  }

  public static GetterSuffixes allUndefined() {
    return new GetterSuffixes(null, null, null, null);
  }

  public GetterSuffixes withCommonSuffixes(GetterSuffixes commonSuffixes) {
    return new GetterSuffixes(
        getRequiredSuffix().orElse(commonSuffixes.getRequiredSuffix().orElse(null)),
        getRequiredNullableSuffix().orElse(commonSuffixes.getRequiredNullableSuffix().orElse(null)),
        getOptionalSuffix().orElse(commonSuffixes.getOptionalSuffix().orElse(null)),
        getOptionalNullableSuffix()
            .orElse(commonSuffixes.getOptionalNullableSuffix().orElse(null)));
  }

  public Optional<String> getRequiredSuffix() {
    return Optional.ofNullable(requiredSuffix);
  }

  public String getRequiredSuffixOrDefault() {
    return getRequiredSuffix().orElse(DEFAULT_REQUIRED);
  }

  public Optional<String> getRequiredNullableSuffix() {
    return Optional.ofNullable(requiredNullableSuffix);
  }

  public String getRequiredNullableSuffixOrDefault() {
    return getRequiredNullableSuffix().orElse(DEFAULT_REQUIRED_NULLABLE);
  }

  public Optional<String> getOptionalSuffix() {
    return Optional.ofNullable(optionalSuffix);
  }

  public String getOptionalSuffixOrDefault() {
    return getOptionalSuffix().orElse(DEFAULT_OPTIONAL);
  }

  public Optional<String> getOptionalNullableSuffix() {
    return Optional.ofNullable(optionalNullableSuffix);
  }

  public String getOptionalNullableSuffixOrDefault() {
    return getOptionalNullableSuffix().orElse(DEFAULT_OPTIONAL_NULLABLE);
  }
}
