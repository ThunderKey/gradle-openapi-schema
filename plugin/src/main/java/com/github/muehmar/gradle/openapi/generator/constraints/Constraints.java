package com.github.muehmar.gradle.openapi.generator.constraints;

import java.util.Objects;
import java.util.Optional;
import java.util.function.Consumer;

public class Constraints {
  private final Min min;
  private final Max max;
  private final DecimalMin decimalMin;
  private final DecimalMax decimalMax;
  private final Size size;
  private final Pattern pattern;
  private final Email email;

  private Constraints(
      Min min,
      Max max,
      DecimalMin decimalMin,
      DecimalMax decimalMax,
      Size size,
      Pattern pattern,
      Email email) {
    this.min = min;
    this.max = max;
    this.decimalMin = decimalMin;
    this.decimalMax = decimalMax;
    this.size = size;
    this.pattern = pattern;
    this.email = email;
  }

  public static Constraints empty() {
    return new Constraints(null, null, null, null, null, null, null);
  }

  public static Constraints ofMin(Min min) {
    return Constraints.empty().withMin(min);
  }

  public static Constraints ofMax(Max max) {
    return Constraints.empty().withMax(max);
  }

  public static Constraints ofMinAndMax(Min min, Max max) {
    return Constraints.empty().withMin(min).withMax(max);
  }

  public static Constraints ofSize(Size size) {
    return Constraints.empty().withSize(size);
  }

  public static Constraints ofPattern(Pattern pattern) {
    return Constraints.empty().withPattern(pattern);
  }

  public static Constraints ofEmail() {
    return Constraints.empty().withEmail(new Email());
  }

  public static Constraints ofDecimalMin(DecimalMin decimalMin) {
    return Constraints.empty().withDecimalMin(decimalMin);
  }

  public static Constraints ofDecimalMax(DecimalMax decimalMax) {
    return Constraints.empty().withDecimalMax(decimalMax);
  }

  public Constraints withMin(Min min) {
    return new Constraints(min, max, decimalMin, decimalMax, size, pattern, email);
  }

  public Constraints withMax(Max max) {
    return new Constraints(min, max, decimalMin, decimalMax, size, pattern, email);
  }

  public Constraints withSize(Size size) {
    return new Constraints(min, max, decimalMin, decimalMax, size, pattern, email);
  }

  public Constraints withPattern(Pattern pattern) {
    return new Constraints(min, max, decimalMin, decimalMax, size, pattern, email);
  }

  public Constraints withEmail(Email email) {
    return new Constraints(min, max, decimalMin, decimalMax, size, pattern, email);
  }

  public Constraints withDecimalMin(DecimalMin decimalMin) {
    return new Constraints(min, max, decimalMin, decimalMax, size, pattern, email);
  }

  public Constraints withDecimalMax(DecimalMax decimalMax) {
    return new Constraints(min, max, decimalMin, decimalMax, size, pattern, email);
  }

  public void onMin(Consumer<Min> onMin) {
    Optional.ofNullable(min).ifPresent(onMin);
  }

  public void onMax(Consumer<Max> onMax) {
    Optional.ofNullable(max).ifPresent(onMax);
  }

  public void onDecimalMin(Consumer<DecimalMin> onDecimalMin) {
    Optional.ofNullable(decimalMin).ifPresent(onDecimalMin);
  }

  public void onDecimalMax(Consumer<DecimalMax> onDecimalMax) {
    Optional.ofNullable(decimalMax).ifPresent(onDecimalMax);
  }

  public void onSize(Consumer<Size> onSize) {
    Optional.ofNullable(size).ifPresent(onSize);
  }

  public void onPattern(Consumer<Pattern> onPattern) {
    Optional.ofNullable(pattern).ifPresent(onPattern);
  }

  public void onEmail(Consumer<Email> onEmail) {
    Optional.ofNullable(email).ifPresent(onEmail);
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    Constraints that = (Constraints) o;
    return Objects.equals(min, that.min)
        && Objects.equals(max, that.max)
        && Objects.equals(size, that.size)
        && Objects.equals(pattern, that.pattern)
        && Objects.equals(email, that.email);
  }

  @Override
  public int hashCode() {
    return Objects.hash(min, max, size, pattern, email);
  }

  @Override
  public String toString() {
    return "Constraints{"
        + "min="
        + min
        + ", max="
        + max
        + ", size="
        + size
        + ", pattern="
        + pattern
        + ", email="
        + email
        + '}';
  }
}
