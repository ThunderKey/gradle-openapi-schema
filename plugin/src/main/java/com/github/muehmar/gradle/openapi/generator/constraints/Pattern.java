package com.github.muehmar.gradle.openapi.generator.constraints;

import java.util.Objects;

public class Pattern {
  private final String value;

  public Pattern(String value) {
    this.value = value;
  }

  public String getPattern() {
    return value;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;
    Pattern pattern = (Pattern) o;
    return Objects.equals(value, pattern.value);
  }

  @Override
  public int hashCode() {
    return Objects.hash(value);
  }

  @Override
  public String toString() {
    return "Pattern{" + "value='" + value + '\'' + '}';
  }
}
