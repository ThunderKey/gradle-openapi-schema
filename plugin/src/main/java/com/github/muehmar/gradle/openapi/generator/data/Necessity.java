package com.github.muehmar.gradle.openapi.generator.data;

/** Defines the necessity of an attribute, i.e. if the attribute is required or optional. */
public enum Necessity {
  REQUIRED,
  OPTIONAL;

  public static Necessity fromBoolean(boolean necessity) {
    return necessity ? REQUIRED : OPTIONAL;
  }
}
