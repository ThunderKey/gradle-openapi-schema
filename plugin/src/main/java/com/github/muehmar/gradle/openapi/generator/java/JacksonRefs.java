package com.github.muehmar.gradle.openapi.generator.java;

public class JacksonRefs {
  private JacksonRefs() {}

  public static final String JSON_POJO_BUILDER =
      "com.fasterxml.jackson.databind.annotation.JsonPOJOBuilder";
}
