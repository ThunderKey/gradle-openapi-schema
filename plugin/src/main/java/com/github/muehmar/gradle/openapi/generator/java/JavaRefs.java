package com.github.muehmar.gradle.openapi.generator.java;

public class JavaRefs {
  private JavaRefs() {}

  public static final String LOCAL_DATE = "java.time.LocalDate";
  public static final String OPTIONAL = "java.util.Optional";
}
