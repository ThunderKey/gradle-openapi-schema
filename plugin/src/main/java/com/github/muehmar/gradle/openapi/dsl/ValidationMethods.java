package com.github.muehmar.gradle.openapi.dsl;

import static com.github.muehmar.gradle.openapi.util.Optionals.or;

import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.generator.settings.JavaModifier;
import io.github.muehmar.pojobuilder.annotations.Nullable;
import io.github.muehmar.pojobuilder.annotations.PojoBuilder;
import java.io.Serializable;
import java.util.Optional;
import lombok.Data;
import org.gradle.api.InvalidUserDataException;

@Data
@PojoBuilder
public class ValidationMethods implements Serializable {
  private static final String DEFAULT_MODIFIER = JavaModifier.PRIVATE.getValue();
  private static final String DEFAULT_SUFFIX = "Raw";
  private static final boolean DEFAULT_DEPRECATED_ANNOTATION = false;

  @Nullable String modifier;
  @Nullable String getterSuffix;
  @Nullable Boolean deprecatedAnnotation;

  public ValidationMethods() {
    this(null, null, null);
  }

  public ValidationMethods(String modifier, String getterSuffix, Boolean deprecatedAnnotation) {
    this.modifier = modifier;
    this.getterSuffix = getterSuffix;
    this.deprecatedAnnotation = deprecatedAnnotation;
  }

  public static ValidationMethods allUndefined() {
    return new ValidationMethods();
  }

  public ValidationMethods withCommonRawGetter(ValidationMethods commonValidationMethods) {
    return ValidationMethodsBuilder.create()
        .andAllOptionals()
        .modifier(or(getModifier(), commonValidationMethods.getModifier()))
        .getterSuffix(or(getGetterSuffix(), commonValidationMethods.getGetterSuffix()))
        .deprecatedAnnotation(
            or(getDeprecatedAnnotation(), commonValidationMethods.getDeprecatedAnnotation()))
        .build();
  }

  public Optional<String> getModifier() {
    return Optional.ofNullable(modifier);
  }

  public JavaModifier getModifierOrDefault() {
    final String modifierString = getModifier().orElse(DEFAULT_MODIFIER);

    return JavaModifier.fromString(modifierString)
        .orElseThrow(
            () ->
                new InvalidUserDataException(
                    "Unsupported value for modifier: '"
                        + modifierString
                        + "'. Supported values are ["
                        + PList.of(JavaModifier.values())
                            .map(JavaModifier::getValue)
                            .mkString(", ")));
  }

  public Optional<String> getGetterSuffix() {
    return Optional.ofNullable(getterSuffix);
  }

  public String getGetterSuffixOrDefault() {
    return getGetterSuffix().orElse(DEFAULT_SUFFIX);
  }

  public Optional<Boolean> getDeprecatedAnnotation() {
    return Optional.ofNullable(deprecatedAnnotation);
  }

  public boolean getDeprecatedAnnotationOrDefault() {
    return getDeprecatedAnnotation().orElse(DEFAULT_DEPRECATED_ANNOTATION);
  }
}
