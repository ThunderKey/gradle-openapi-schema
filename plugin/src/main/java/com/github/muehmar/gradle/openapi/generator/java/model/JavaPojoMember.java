package com.github.muehmar.gradle.openapi.generator.java.model;

import com.github.muehmar.gradle.openapi.generator.java.model.type.JavaType;
import com.github.muehmar.gradle.openapi.generator.model.Name;
import com.github.muehmar.gradle.openapi.generator.model.Necessity;
import com.github.muehmar.gradle.openapi.generator.model.Nullability;
import com.github.muehmar.gradle.openapi.generator.model.PojoMember;
import com.github.muehmar.gradle.openapi.generator.settings.GetterSuffixes;
import com.github.muehmar.gradle.openapi.generator.settings.PojoSettings;
import com.github.muehmar.gradle.openapi.generator.settings.TypeMappings;
import lombok.EqualsAndHashCode;
import lombok.ToString;

@EqualsAndHashCode
@ToString
public class JavaPojoMember {
  private final Name name;
  private final String description;
  private final JavaType javaType;
  private final Necessity necessity;
  private final Nullability nullability;

  private JavaPojoMember(
      Name name,
      String description,
      JavaType javaType,
      Necessity necessity,
      Nullability nullability) {
    this.javaType = javaType;
    this.name = name;
    this.description = description;
    this.necessity = necessity;
    this.nullability = nullability;
  }

  public static JavaPojoMember of(
      Name name,
      String description,
      JavaType javaType,
      Necessity necessity,
      Nullability nullability) {
    return new JavaPojoMember(name, description, javaType, necessity, nullability);
  }

  public static JavaPojoMember wrap(PojoMember pojoMember, TypeMappings typeMappings) {
    final JavaType javaType = JavaType.wrap(pojoMember.getType(), typeMappings);
    return new JavaPojoMember(
        pojoMember.getName(),
        pojoMember.getDescription(),
        javaType,
        pojoMember.getNecessity(),
        pojoMember.getNullability());
  }

  public Nullability getNullability() {
    return nullability;
  }

  public Necessity getNecessity() {
    return necessity;
  }

  public String getDescription() {
    return description;
  }

  public Name getName() {
    return name;
  }

  public JavaType getJavaType() {
    return javaType;
  }

  public boolean isOptional() {
    return necessity.isOptional();
  }

  public boolean isRequired() {
    return necessity.isRequired();
  }

  public boolean isNullable() {
    return nullability.isNullable();
  }

  public boolean isNotNullable() {
    return nullability.isNotNullable();
  }

  public boolean isRequiredAndNullable() {
    return isRequired() && isNullable();
  }

  public boolean isRequiredAndNotNullable() {
    return isRequired() && isNotNullable();
  }

  public boolean isOptionalAndNullable() {
    return isOptional() && isNullable();
  }

  public boolean isOptionalAndNotNullable() {
    return isOptional() && isNotNullable();
  }

  public Name getGetterName() {
    return prefixedMethodName("get");
  }

  public Name getGetterNameWithSuffix(PojoSettings settings) {
    return getGetterName().append(determineSuffix(settings));
  }

  private String determineSuffix(PojoSettings settings) {
    final GetterSuffixes getterSuffixes = settings.getGetterSuffixes();
    if (isRequiredAndNotNullable()) {
      return getterSuffixes.getRequiredSuffix();
    } else if (isRequiredAndNullable()) {
      return getterSuffixes.getRequiredNullableSuffix();
    } else if (isOptionalAndNotNullable()) {
      return getterSuffixes.getOptionalSuffix();
    } else {
      return getterSuffixes.getOptionalNullableSuffix();
    }
  }

  public Name getWitherName() {
    return prefixedMethodName("with");
  }

  public Name prefixedMethodName(String prefix) {
    if (prefix.isEmpty()) {
      return name;
    } else {
      return name.startUpperCase().prefix(prefix);
    }
  }
}
