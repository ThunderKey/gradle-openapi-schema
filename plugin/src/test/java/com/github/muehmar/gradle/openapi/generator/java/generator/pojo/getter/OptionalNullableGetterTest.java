package com.github.muehmar.gradle.openapi.generator.java.generator.pojo.getter;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.junit5.SnapshotExtension;
import com.github.muehmar.gradle.openapi.generator.java.JacksonRefs;
import com.github.muehmar.gradle.openapi.generator.java.JavaRefs;
import com.github.muehmar.gradle.openapi.generator.java.OpenApiUtilRefs;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaPojoMember;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaPojoMembers;
import com.github.muehmar.gradle.openapi.generator.model.Necessity;
import com.github.muehmar.gradle.openapi.generator.model.Nullability;
import com.github.muehmar.gradle.openapi.generator.model.constraints.Constraints;
import com.github.muehmar.gradle.openapi.generator.model.constraints.Pattern;
import com.github.muehmar.gradle.openapi.generator.settings.GetterSuffixes;
import com.github.muehmar.gradle.openapi.generator.settings.GetterSuffixesBuilder;
import com.github.muehmar.gradle.openapi.generator.settings.JavaModifier;
import com.github.muehmar.gradle.openapi.generator.settings.JsonSupport;
import com.github.muehmar.gradle.openapi.generator.settings.PojoSettings;
import com.github.muehmar.gradle.openapi.generator.settings.TestPojoSettings;
import com.github.muehmar.gradle.openapi.generator.settings.ValidationMethods;
import io.github.muehmar.codegenerator.Generator;
import io.github.muehmar.codegenerator.writer.Writer;
import java.util.function.Function;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;

@ExtendWith(SnapshotExtension.class)
class OptionalNullableGetterTest {

  private Expect expect;

  @Test
  void generator_when_enabledJacksonAndDisabledValidation_then_correctOutputAndRefs() {
    final Generator<JavaPojoMember, PojoSettings> generator = OptionalNullableGetter.getter();
    final JavaPojoMember pojoMember =
        JavaPojoMembers.birthdate(Necessity.OPTIONAL, Nullability.NULLABLE);

    final Writer writer =
        generator.generate(
            pojoMember,
            TestPojoSettings.defaultSettings().withEnableValidation(false),
            Writer.createDefault());

    assertEquals(
        6,
        writer.getRefs().distinct(Function.identity()).size(),
        "Refs: " + writer.getRefs().mkString(", "));
    assertTrue(writer.getRefs().exists(JavaRefs.JAVA_TIME_LOCAL_DATE::equals));
    assertTrue(writer.getRefs().exists(OpenApiUtilRefs.TRISTATE::equals));
    assertTrue(writer.getRefs().exists(OpenApiUtilRefs.JACKSON_NULL_CONTAINER::equals));
    assertTrue(writer.getRefs().exists(JacksonRefs.JSON_IGNORE::equals));
    assertTrue(writer.getRefs().exists(JacksonRefs.JSON_PROPERTY::equals));
    assertTrue(writer.getRefs().exists(JacksonRefs.JSON_INCLUDE::equals));

    expect.toMatchSnapshot(writer.asString());
  }

  @Test
  void generator_when_disabledJacksonAndEnabledValidation_then_correctOutputAndRefs() {
    final Generator<JavaPojoMember, PojoSettings> generator = OptionalNullableGetter.getter();
    final JavaPojoMember pojoMember =
        JavaPojoMembers.birthdate(
            Constraints.ofPattern(Pattern.ofUnescapedString("DatePattern")),
            Necessity.OPTIONAL,
            Nullability.NULLABLE);

    final Writer writer =
        generator.generate(
            pojoMember,
            TestPojoSettings.defaultSettings().withJsonSupport(JsonSupport.NONE),
            Writer.createDefault());

    assertTrue(writer.getRefs().exists(JavaRefs.JAVA_TIME_LOCAL_DATE::equals));
    assertTrue(writer.getRefs().exists(OpenApiUtilRefs.TRISTATE::equals));

    expect.toMatchSnapshot(writer.asString());
  }

  @Test
  void generator_when_optionalNullableSuffix_then_correctOutputAndRefs() {
    final Generator<JavaPojoMember, PojoSettings> generator = OptionalNullableGetter.getter();
    final JavaPojoMember pojoMember =
        JavaPojoMembers.birthdate(
            Constraints.ofPattern(Pattern.ofUnescapedString("DatePattern")),
            Necessity.OPTIONAL,
            Nullability.NULLABLE);

    final GetterSuffixes getterSuffixes =
        GetterSuffixesBuilder.create()
            .requiredSuffix("")
            .requiredNullableSuffix("")
            .optionalSuffix("")
            .optionalNullableSuffix("OptNull")
            .build();

    final Writer writer =
        generator.generate(
            pojoMember,
            TestPojoSettings.defaultSettings()
                .withJsonSupport(JsonSupport.NONE)
                .withEnableValidation(false)
                .withGetterSuffixes(getterSuffixes),
            Writer.createDefault());

    assertTrue(writer.getRefs().exists(JavaRefs.JAVA_TIME_LOCAL_DATE::equals));
    assertTrue(writer.getRefs().exists(OpenApiUtilRefs.TRISTATE::equals));

    expect.toMatchSnapshot(writer.asString());
  }

  @Test
  void generator_when_deprecatedAnnotation_then_correctOutputAndRefs() {
    final Generator<JavaPojoMember, PojoSettings> generator = OptionalNullableGetter.getter();
    final JavaPojoMember pojoMember =
        JavaPojoMembers.birthdate(Necessity.OPTIONAL, Nullability.NULLABLE);

    final ValidationMethods validationMethods =
        TestPojoSettings.defaultValidationMethods()
            .withDeprecatedAnnotation(true)
            .withModifier(JavaModifier.PUBLIC);

    final Writer writer =
        generator.generate(
            pojoMember,
            TestPojoSettings.defaultSettings()
                .withJsonSupport(JsonSupport.JACKSON)
                .withValidationMethods(validationMethods),
            Writer.createDefault());

    expect.toMatchSnapshot(writer.asString());
  }
}
