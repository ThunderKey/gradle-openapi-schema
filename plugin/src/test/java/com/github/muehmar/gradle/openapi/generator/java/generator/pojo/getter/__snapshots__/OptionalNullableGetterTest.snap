com.github.muehmar.gradle.openapi.generator.java.generator.pojo.getter.OptionalNullableGetterTest.generator_when_deprecatedAnnotation_then_correctOutputAndRefs=[
com.fasterxml.jackson.annotation.JsonIgnore
com.fasterxml.jackson.annotation.JsonInclude
com.fasterxml.jackson.annotation.JsonProperty
com.github.muehmar.openapi.util.JacksonNullContainer
com.github.muehmar.openapi.util.Tristate
java.time.LocalDate

/**
 * Birthdate
 */
@JsonIgnore
public Tristate<LocalDate> getBirthdate() {
  return Tristate.ofNullableAndNullFlag(birthdate, isBirthdateNull);
}

@JsonProperty("birthdate")
@JsonInclude(JsonInclude.Include.NON_NULL)
private Object getBirthdateJackson() {
  return isBirthdateNull ? new JacksonNullContainer<>(birthdate) : birthdate;
}

/**
 * @deprecated This method is intended to be used only by frameworks for
 * validation or serialisation.
 */
@Deprecated
public LocalDate getBirthdateRaw() {
  return birthdate;
}
]


com.github.muehmar.gradle.openapi.generator.java.generator.pojo.getter.OptionalNullableGetterTest.generator_when_disabledJacksonAndEnabledValidation_then_correctOutputAndRefs=[
com.github.muehmar.openapi.util.Tristate
java.time.LocalDate
javax.validation.constraints.Pattern

/**
 * Birthdate
 */
public Tristate<LocalDate> getBirthdate() {
  return Tristate.ofNullableAndNullFlag(birthdate, isBirthdateNull);
}

@Pattern(regexp="DatePattern")
private LocalDate getBirthdateRaw() {
  return birthdate;
}
]


com.github.muehmar.gradle.openapi.generator.java.generator.pojo.getter.OptionalNullableGetterTest.generator_when_enabledJacksonAndDisabledValidation_then_correctOutputAndRefs=[
com.fasterxml.jackson.annotation.JsonIgnore
com.fasterxml.jackson.annotation.JsonInclude
com.fasterxml.jackson.annotation.JsonProperty
com.github.muehmar.openapi.util.JacksonNullContainer
com.github.muehmar.openapi.util.Tristate
java.time.LocalDate

/**
 * Birthdate
 */
@JsonIgnore
public Tristate<LocalDate> getBirthdate() {
  return Tristate.ofNullableAndNullFlag(birthdate, isBirthdateNull);
}

@JsonProperty("birthdate")
@JsonInclude(JsonInclude.Include.NON_NULL)
private Object getBirthdateJackson() {
  return isBirthdateNull ? new JacksonNullContainer<>(birthdate) : birthdate;
}
]


com.github.muehmar.gradle.openapi.generator.java.generator.pojo.getter.OptionalNullableGetterTest.generator_when_optionalNullableSuffix_then_correctOutputAndRefs=[
com.github.muehmar.openapi.util.Tristate
java.time.LocalDate

/**
 * Birthdate
 */
public Tristate<LocalDate> getBirthdateOptNull() {
  return Tristate.ofNullableAndNullFlag(birthdate, isBirthdateNull);
}
]